src/VisualStudio/Core/Def/Implementation/VirtualMemoryNotificationListener.cs

```diff
@@ -104,5 +116,46 @@ public int OnBroadcastMessage(uint msg, IntPtr wParam, IntPtr lParam)
 
             return VSConstants.S_OK;
         }
+
+        private bool ShouldTurnOffFullSolutionAnalysis(long availableMemory)
+        {
+            // conditions
+            // 1. if available memory is less than the threshold and 
+            // 2. if full solution analysis memory monitor is on (user can set it off using registery, when he does, we will never show info bar) and
+            // 3. if our full solution analysis option is on (not user full solution analysis option, but our internal one) and
+            // 4. if infobar is never shown to users for this solution
+            return availableMemory < MemoryThreshold &&
+                  _workspace.Options.GetOption(InternalFeatureOnOffOptions.FullSolutionAnalysisMemoryMonitor) &&
+                  _workspace.Options.GetOption(RuntimeOptions.FullSolutionAnalysis) &&
+                  !_workspace.Options.GetOption(RuntimeOptions.FullSolutionAnalysisInfoBarShown);
+        }
+
+        private bool IsUserOptionOn()
+        {
+            // check languages currently on solution. since we only show info bar once, we don't need to track solution changes.
+            var languages = _workspace.CurrentSolution.Projects.Select(p => p.Language).Distinct();
+            foreach (var language in languages)
+            {
+                if (ServiceFeatureOnOffOptions.IsClosedFileDiagnosticsEnabled(_workspace, language))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            if (e.Kind != WorkspaceChangeKind.SolutionAdded)
+            {
+                return;
+            }
+
+            // first make sure full solution analysis is on. (not user options but our internal options. even if our option is on, if user option is off
```

Comment:
here are some feedback from users.

https://github.com/dotnet/roslyn/issues/7082#issuecomment-221003598
"would it be possible to make some of these options of disabling these tools available per solution? We have one legacy solution that simply brings VS to it's knees with codelens and full solution analysis enabled. We use it maybe 5% of our development time but most of my team has turned those tools off across the board. I've taken to creating an entire new hive of VS that I launch just for that project which seems to be the "best" approach in terms of keeping VS usable for this one project for me, but a tough one to remember to do for the rest of my team."